### DISCUSSION TITLE
Automatic Authentication

### DISCUSSION THREAD

**Discussion by samjohnson1357** [jessecooper/pyetrade#64] [Ideas]
Normally you have to go to the E*TRADE webpage to obtain a code and authenticate.  What ways do you guys have of automating this process?

**Comment by Token-DAO** [comment]
I'm still trying to figure out how to automate authentication with what E*Trade provides in the [documentation](https://developer.etrade.com/getting-started/developer-guides). I'm trying to do it in Jupyter Notebook with Python. 

> Callbacks
>
> As mentioned above, when the user authorizes the application, the E*TRADE website generates a verification code that must be passed to the application. One approach is for the user to simply copy the code and paste it into the application. A much better solution is for E*TRADE to automatically redirect the user back to the application, using a callback URL with the verification code added as a query parameter, as shown in these example URLs:

>
> https://myapplicationsite.com/mytradingapp?oauth_verifier=WXYZ89
> 
> [https://myapplicationsite.com?myapp=trading&oauth_verifier=WXYZ89](https://myapplicationsite.com/?myapp=trading&oauth_verifier=WXYZ89)
> 
> Configuring a callback
> 
> Using a callback requires that the callback URL be associated with your consumer key in the E*TRADE system. To request this, log in to your E*TRADE account and send a secure message to Customer Service. Select the subject "Technical Issues" and the topic "E*TRADE API". State that you would like to have a callback configured, and specify your consumer key and the desired callback URL. Your callback URL can be just a simple address, or can also include query parameters.
> 
> Once the callback is configured, two system behaviors are changed:
> 
> The oauth_callback_confirmed property of the request_token API returns TRUE to show that there is a callback URL associated with the consumer key.
> 
> Users who approve the authorization request are automatically redirected to the callback URL, with the verification code appended as a query parameter (as shown in the example URLs above).

**Reply by nathanramoscfa** [reply]
Figured it out: https://github.com/nathanramoscfa/pyetrade

**Reply by egelberg** [reply]
@nathanramoscfa Which cookie are you referring to in [these lines](https://github.com/nathanramoscfa/pyetrade/blob/master/pyetrade/authorization.py#L28-L29)? If I were to contact the help desk what would I be asking for?

**Reply by nathanramoscfa** [reply]
@egelberg This [documentation ](https://etradebot.readthedocs.io/en/latest/credentials.html)provides some guidance. Contact E-Trade Customer Service by phone. Best call times might be around 8-9AM EST. Tell them you are a developer with a Python application to automate trading and that you need an authentication cookie for your application. They'll walk you through obtaining the authentication cookie, which may include them reaching out to the E-Trade API team. This might take a day or two. You'll receive the authentication cookie by secure message within the E-Trade website messages feature, and it will be a Python dictionary, generic example below:

```
{
    'name': '<COOKIE_NAME>',
    'value': '<COOKE_VALUE>',
    'domain': '.etrade.com',
    'secure': True,
    'httpOnly': True
}
```

Take a look at this [example](https://github.com/nathanramoscfa/pyetrade#example-usage). You would simply copy and paste the authentication cookie into the variable called "swh_cookie".

**Reply by githubtar** [reply]
Still works?

**Comment by mw66** [comment]
You can get automatedTrading.py here:

https://github.com/jessecooper/pyetrade/pull/65

**Reply by nathanramoscfa** [reply]
automatedTrading.py [no longer works](https://stackoverflow.com/questions/69875125/find-element-by-commands-are-deprecated-in-selenium) with the latest version of Selenium. I forked pyetrade (https://github.com/nathanramoscfa/pyetrade) and added an automated login function that is compatible with Selenium version 4.5.0. It may also be compatible with more recent versions of Selenium but I haven't tested it.
